src/pgwire/src/protocol.rs

```diff
@@ -1232,9 +1232,11 @@ where
                     total_sent_rows += drain_rows;
                     want_rows -= drain_rows;
                     // If we have sent the number of requested rows, put the remainder of the batch
-                    // back and stop sending.
-                    if want_rows == 0 && !batch_rows.is_empty() {
-                        rows = Box::new(stream::iter(vec![batch_rows]).chain(rows));
+                    // (if any) back and stop sending.
+                    if want_rows == 0 {
+                        if !batch_rows.is_empty() {
+                            rows = Box::new(stream::iter(vec![batch_rows]).chain(rows));
+                        }
                         break;
                     }
                     self.conn.flush().await?;
```

Comment:
Do we care about this `conn.flush()` not getting called if we hit the break? I imagine it naturally flushes as a consequence of the command returning, but I wanted to confirm

Reply:
Hmm I dunno this part of the codebase well enough to say for sure. It seems like your change keeps the behavior the same though which is nice.

@mjibson I think you're probably best qualified to answer the original question?

Reply:
Hi @mjibson! :wave:

It does differ a touch in the exact case we're looking at here. If there's a timeout but sufficient messages are immediately available, then we used to flush and wait the timeout but now we do neither. Anyway, I think the flush should always happen via the `self.ready()` call at the end of `StateMachine::query`.

https://github.com/MaterializeInc/materialize/blob/199830cd023585a8a86afdb9458ad388d17fb745/src/pgwire/src/protocol.rs#L448

Reply:
I can't come up with a reason why we need this flush at all. I commented it out and the FETCH and TAIL tests still seem to pass? So I think this is fine.

Reply:
Does FLUSH stream results back? I'd assumed this was for the case where 1 and 2 show up, then we have to wait a bit, then 3 shows up. This would get 1 and 2 back to the client before waiting on 3. Maybe no-one really uses FLUSH like that though

Reply:
Yes, it would stream results. I forgot about streaming TAIL (not using FETCH) which would need this flush. So I think it's fine as is because the new case here doesn't care about the flush.